File: electron/database/paperBoardRepository.ts
import type Database from "better-sqlite3";
import type { PaperBoard, PaperNode, PaperEdge } from "../types";

// --- Row types (snake_case) ---

interface BoardRow {
  id: string;
  name: string;
  linked_note_id: string | null;
  viewport_x: number;
  viewport_y: number;
  viewport_zoom: number;
  order: number;
  created_at: string;
  updated_at: string;
}

interface NodeRow {
  id: string;
  board_id: string;
  node_type: string;
  position_x: number;
  position_y: number;
  width: number;
  height: number;
  z_index: number;
  parent_node_id: string | null;
  ref_entity_id: string | null;
  ref_entity_type: string | null;
  text_content: string | null;
  frame_color: string | null;
  frame_label: string | null;
  created_at: string;
  updated_at: string;
}

interface EdgeRow {
  id: string;
  board_id: string;
  source_node_id: string;
  target_node_id: string;
  source_handle: string | null;
  target_handle: string | null;
  label: string | null;
  style_json: string | null;
  created_at: string;
}

// --- Row → Domain converters ---

function rowToBoard(row: BoardRow): PaperBoard {
  return {
    id: row.id,
    name: row.name,
    linkedNoteId: row.linked_note_id,
    viewportX: row.viewport_x,
    viewportY: row.viewport_y,
    viewportZoom: row.viewport_zoom,
    order: row.order,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

function rowToNode(row: NodeRow): PaperNode {
  return {
    id: row.id,
    boardId: row.board_id,
    nodeType: row.node_type as PaperNode["nodeType"],
    positionX: row.position_x,
    positionY: row.position_y,
    width: row.width,
    height: row.height,
    zIndex: row.z_index,
    parentNodeId: row.parent_node_id,
    refEntityId: row.ref_entity_id,
    refEntityType: row.ref_entity_type,
    textContent: row.text_content,
    frameColor: row.frame_color,
    frameLabel: row.frame_label,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

function rowToEdge(row: EdgeRow): PaperEdge {
  return {
    id: row.id,
    boardId: row.board_id,
    sourceNodeId: row.source_node_id,
    targetNodeId: row.target_node_id,
    sourceHandle: row.source_handle,
    targetHandle: row.target_handle,
    label: row.label,
    styleJson: row.style_json,
    createdAt: row.created_at,
  };
}

// --- Repository factory ---

export function createPaperBoardRepository(db: Database.Database) {
  const stmts = {
    // Boards
    fetchAllBoards: db.prepare(
      `SELECT * FROM paper_boards ORDER BY "order" ASC, created_at ASC`,
    ),
    fetchBoardById: db.prepare(`SELECT * FROM paper_boards WHERE id = ?`),
    fetchBoardByNoteId: db.prepare(
      `SELECT * FROM paper_boards WHERE linked_note_id = ?`,
    ),
    insertBoard: db.prepare(`
      INSERT INTO paper_boards (id, name, linked_note_id, viewport_x, viewport_y, viewport_zoom, "order", created_at, updated_at)
      VALUES (@id, @name, @linked_note_id, @viewport_x, @viewport_y, @viewport_zoom, @order, @created_at, @updated_at)
    `),
    updateBoard: db.prepare(`
      UPDATE paper_boards
      SET name = @name, linked_note_id = @linked_note_id,
          viewport_x = @viewport_x, viewport_y = @viewport_y, viewport_zoom = @viewport_zoom,
          "order" = @order, updated_at = @updated_at
      WHERE id = @id
    `),
    deleteBoard: db.prepare(`DELETE FROM paper_boards WHERE id = ?`),

    // Nodes
    fetchNodesByBoard: db.prepare(
      `SELECT * FROM paper_nodes WHERE board_id = ? ORDER BY z_index ASC`,
    ),
    fetchNodeById: db.prepare(`SELECT * FROM paper_nodes WHERE id = ?`),
    insertNode: db.prepare(`
      INSERT INTO paper_nodes (id, board_id, node_type, position_x, position_y, width, height, z_index, parent_node_id, ref_entity_id, ref_entity_type, text_content, frame_color, frame_label, created_at, updated_at)
      VALUES (@id, @board_id, @node_type, @position_x, @position_y, @width, @height, @z_index, @parent_node_id, @ref_entity_id, @ref_entity_type, @text_content, @frame_color, @frame_label, @created_at, @updated_at)
    `),
    updateNode: db.prepare(`
      UPDATE paper_nodes
      SET position_x = @position_x, position_y = @position_y, width = @width, height = @height,
          z_index = @z_index, parent_node_id = @parent_node_id,
          text_content = @text_content, frame_color = @frame_color, frame_label = @frame_label,
          updated_at = @updated_at
      WHERE id = @id
    `),
    updateNodePosition: db.prepare(`
      UPDATE paper_nodes SET position_x = @position_x, position_y = @position_y, parent_node_id = @parent_node_id, updated_at = @updated_at WHERE id = @id
    `),
    deleteNode: db.prepare(`DELETE FROM paper_nodes WHERE id = ?`),

    // Edges
    fetchEdgesByBoard: db.prepare(
      `SELECT * FROM paper_edges WHERE board_id = ? ORDER BY created_at ASC`,
    ),
    insertEdge: db.prepare(`
      INSERT INTO paper_edges (id, board_id, source_node_id, target_node_id, source_handle, target_handle, label, style_json, created_at)
      VALUES (@id, @board_id, @source_node_id, @target_node_id, @source_handle, @target_handle, @label, @style_json, @created_at)
    `),
    deleteEdge: db.prepare(`DELETE FROM paper_edges WHERE id = ?`),
  };

  const bulkUpdatePositionsTx = db.transaction(
    (
      updates: Array<{
        id: string;
        positionX: number;
        positionY: number;
        parentNodeId: string | null;
      }>,
    ) => {
      const now = new Date().toISOString();
      for (const u of updates) {
        stmts.updateNodePosition.run({
          id: u.id,
          position_x: u.positionX,
          position_y: u.positionY,
          parent_node_id: u.parentNodeId,
          updated_at: now,
        });
      }
    },
  );

  return {
    // --- Boards ---
    fetchAllBoards(): PaperBoard[] {
      return (stmts.fetchAllBoards.all() as BoardRow[]).map(rowToBoard);
    },

    fetchBoardById(id: string): PaperBoard | null {
      const row = stmts.fetchBoardById.get(id) as BoardRow | undefined;
      return row ? rowToBoard(row) : null;
    },

    fetchBoardByNoteId(noteId: string): PaperBoard | null {
      const row = stmts.fetchBoardByNoteId.get(noteId) as BoardRow | undefined;
      return row ? rowToBoard(row) : null;
    },

    createBoard(name: string, linkedNoteId?: string | null): PaperBoard {
      const now = new Date().toISOString();
      const id = `pb-${crypto.randomUUID()}`;
      const maxOrder =
        (
          db.prepare(`SELECT MAX("order") as m FROM paper_boards`).get() as {
            m: number | null;
          }
        ).m ?? -1;
      stmts.insertBoard.run({
        id,
        name,
        linked_note_id: linkedNoteId ?? null,
        viewport_x: 0,
        viewport_y: 0,
        viewport_zoom: 1,
        order: maxOrder + 1,
        created_at: now,
        updated_at: now,
      });
      return rowToBoard(stmts.fetchBoardById.get(id) as BoardRow);
    },

    updateBoard(
      id: string,
      updates: Partial<
        Pick<
          PaperBoard,
          | "name"
          | "linkedNoteId"
          | "viewportX"
          | "viewportY"
          | "viewportZoom"
          | "order"
        >
      >,
    ): PaperBoard {
      const existing = stmts.fetchBoardById.get(id) as BoardRow;
      if (!existing) throw new Error(`Board not found: ${id}`);
      const now = new Date().toISOString();
      stmts.updateBoard.run({
        id,
        name: updates.name ?? existing.name,
        linked_note_id:
          updates.linkedNoteId !== undefined
            ? updates.linkedNoteId
            : existing.linked_note_id,
        viewport_x: updates.viewportX ?? existing.viewport_x,
        viewport_y: updates.viewportY ?? existing.viewport_y,
        viewport_zoom: updates.viewportZoom ?? existing.viewport_zoom,
        order: updates.order ?? existing.order,
        updated_at: now,
      });
      return rowToBoard(stmts.fetchBoardById.get(id) as BoardRow);
    },

    deleteBoard(id: string): void {
      stmts.deleteBoard.run(id);
    },

    // --- Nodes ---
    fetchNodesByBoard(boardId: string): PaperNode[] {
      return (stmts.fetchNodesByBoard.all(boardId) as NodeRow[]).map(rowToNode);
    },

    createNode(params: {
      boardId: string;
      nodeType: PaperNode["nodeType"];
      positionX: number;
      positionY: number;
      width?: number;
      height?: number;
      zIndex?: number;
      parentNodeId?: string | null;
      refEntityId?: string | null;
      refEntityType?: string | null;
      textContent?: string | null;
      frameColor?: string | null;
      frameLabel?: string | null;
    }): PaperNode {
      const now = new Date().toISOString();
      const id = `pn-${crypto.randomUUID()}`;
      stmts.insertNode.run({
        id,
        board_id: params.boardId,
        node_type: params.nodeType,
        position_x: params.positionX,
        position_y: params.positionY,
        width: params.width ?? 200,
        height: params.height ?? 100,
        z_index: params.zIndex ?? 0,
        parent_node_id: params.parentNodeId ?? null,
        ref_entity_id: params.refEntityId ?? null,
        ref_entity_type: params.refEntityType ?? null,
        text_content: params.textContent ?? null,
        frame_color: params.frameColor ?? null,
        frame_label: params.frameLabel ?? null,
        created_at: now,
        updated_at: now,
      });
      return rowToNode(stmts.fetchNodeById.get(id) as NodeRow);
    },

    updateNode(
      id: string,
      updates: Partial<
        Pick<
          PaperNode,
          | "positionX"
          | "positionY"
          | "width"
          | "height"
          | "zIndex"
          | "parentNodeId"
          | "textContent"
          | "frameColor"
          | "frameLabel"
        >
      >,
    ): PaperNode {
      const existing = stmts.fetchNodeById.get(id) as NodeRow;
      if (!existing) throw new Error(`Node not found: ${id}`);
      const now = new Date().toISOString();
      stmts.updateNode.run({
        id,
        position_x: updates.positionX ?? existing.position_x,
        position_y: updates.positionY ?? existing.position_y,
        width: updates.width ?? existing.width,
        height: updates.height ?? existing.height,
        z_index: updates.zIndex ?? existing.z_index,
        parent_node_id:
          updates.parentNodeId !== undefined
            ? updates.parentNodeId
            : existing.parent_node_id,
        text_content:
          updates.textContent !== undefined
            ? updates.textContent
            : existing.text_content,
        frame_color:
          updates.frameColor !== undefined
            ? updates.frameColor
            : existing.frame_color,
        frame_label:
          updates.frameLabel !== undefined
            ? updates.frameLabel
            : existing.frame_label,
        updated_at: now,
      });
      return rowToNode(stmts.fetchNodeById.get(id) as NodeRow);
    },

    bulkUpdatePositions(
      updates: Array<{
        id: string;
        positionX: number;
        positionY: number;
        parentNodeId: string | null;
      }>,
    ): void {
      bulkUpdatePositionsTx(updates);
    },

    deleteNode(id: string): void {
      stmts.deleteNode.run(id);
    },

    // --- Edges ---
    fetchEdgesByBoard(boardId: string): PaperEdge[] {
      return (stmts.fetchEdgesByBoard.all(boardId) as EdgeRow[]).map(rowToEdge);
    },

    createEdge(params: {
      boardId: string;
      sourceNodeId: string;
      targetNodeId: string;
      sourceHandle?: string | null;
      targetHandle?: string | null;
      label?: string | null;
      styleJson?: string | null;
    }): PaperEdge {
      const now = new Date().toISOString();
      const id = `pe-${crypto.randomUUID()}`;
      stmts.insertEdge.run({
        id,
        board_id: params.boardId,
        source_node_id: params.sourceNodeId,
        target_node_id: params.targetNodeId,
        source_handle: params.sourceHandle ?? null,
        target_handle: params.targetHandle ?? null,
        label: params.label ?? null,
        style_json: params.styleJson ?? null,
        created_at: now,
      });
      return rowToEdge(
        stmts.fetchEdgesByBoard
          .all(params.boardId)
          .find((r: any) => r.id === id) as EdgeRow,
      );
    },

    deleteEdge(id: string): void {
      stmts.deleteEdge.run(id);
    },
  };
}

export type PaperBoardRepository = ReturnType<
  typeof createPaperBoardRepository
>;

File: src-tauri/src/db/paper_board_repository.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::helpers;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PaperBoard {
    pub id: String,
    pub name: String,
    pub linked_note_id: Option<String>,
    pub viewport_x: f64,
    pub viewport_y: f64,
    pub viewport_zoom: f64,
    pub order: i64,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PaperNode {
    pub id: String,
    pub board_id: String,
    pub node_type: String,
    pub position_x: f64,
    pub position_y: f64,
    pub width: f64,
    pub height: f64,
    pub z_index: i64,
    pub parent_node_id: Option<String>,
    pub ref_entity_id: Option<String>,
    pub ref_entity_type: Option<String>,
    pub text_content: Option<String>,
    pub frame_color: Option<String>,
    pub frame_label: Option<String>,
    pub label: Option<String>,
    pub hidden: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PaperEdge {
    pub id: String,
    pub board_id: String,
    pub source_node_id: String,
    pub target_node_id: String,
    pub source_handle: Option<String>,
    pub target_handle: Option<String>,
    pub label: Option<String>,
    pub style_json: Option<String>,
    pub created_at: String,
}

fn row_to_board(row: &rusqlite::Row) -> rusqlite::Result<PaperBoard> {
    Ok(PaperBoard {
        id: row.get("id")?,
        name: row.get("name")?,
        linked_note_id: row.get("linked_note_id")?,
        viewport_x: row.get("viewport_x")?,
        viewport_y: row.get("viewport_y")?,
        viewport_zoom: row.get("viewport_zoom")?,
        order: row.get("\"order\"").or_else(|_| row.get("order"))?,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
    })
}

fn row_to_node(row: &rusqlite::Row) -> rusqlite::Result<PaperNode> {
    Ok(PaperNode {
        id: row.get("id")?,
        board_id: row.get("board_id")?,
        node_type: row.get("node_type")?,
        position_x: row.get("position_x")?,
        position_y: row.get("position_y")?,
        width: row.get("width")?,
        height: row.get("height")?,
        z_index: row.get("z_index")?,
        parent_node_id: row.get("parent_node_id")?,
        ref_entity_id: row.get("ref_entity_id")?,
        ref_entity_type: row.get("ref_entity_type")?,
        text_content: row.get("text_content")?,
        frame_color: row.get("frame_color")?,
        frame_label: row.get("frame_label")?,
        label: row.get("label")?,
        hidden: row.get::<_, i64>("hidden")? != 0,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
    })
}

fn row_to_edge(row: &rusqlite::Row) -> rusqlite::Result<PaperEdge> {
    Ok(PaperEdge {
        id: row.get("id")?,
        board_id: row.get("board_id")?,
        source_node_id: row.get("source_node_id")?,
        target_node_id: row.get("target_node_id")?,
        source_handle: row.get("source_handle")?,
        target_handle: row.get("target_handle")?,
        label: row.get("label")?,
        style_json: row.get("style_json")?,
        created_at: row.get("created_at")?,
    })
}

// --- Board functions ---

pub fn fetch_all_boards(conn: &Connection) -> rusqlite::Result<Vec<PaperBoard>> {
    let mut stmt =
        conn.prepare("SELECT * FROM paper_boards ORDER BY \"order\" ASC")?;
    let rows = stmt.query_map([], |row| row_to_board(row))?;
    rows.collect()
}

pub fn fetch_board_by_id(
    conn: &Connection,
    id: &str,
) -> rusqlite::Result<Option<PaperBoard>> {
    let mut stmt = conn.prepare("SELECT * FROM paper_boards WHERE id = ?1")?;
    let result = stmt.query_row([id], |row| row_to_board(row));
    match result {
        Ok(board) => Ok(Some(board)),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(e),
    }
}

pub fn fetch_board_by_note_id(
    conn: &Connection,
    note_id: &str,
) -> rusqlite::Result<Option<PaperBoard>> {
    let mut stmt =
        conn.prepare("SELECT * FROM paper_boards WHERE linked_note_id = ?1")?;
    let result = stmt.query_row([note_id], |row| row_to_board(row));
    match result {
        Ok(board) => Ok(Some(board)),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(e),
    }
}

pub fn create_board(
    conn: &Connection,
    name: &str,
    linked_note_id: Option<&str>,
) -> rusqlite::Result<PaperBoard> {
    let id = format!("board-{}", helpers::new_uuid());
    let now = helpers::now();
    let order = helpers::next_order(conn, "paper_boards")?;

    conn.execute(
        "INSERT INTO paper_boards \
         (id, name, linked_note_id, viewport_x, viewport_y, viewport_zoom, \
          \"order\", created_at, updated_at) \
         VALUES (?1, ?2, ?3, 0.0, 0.0, 1.0, ?4, ?5, ?6)",
        params![&id, name, linked_note_id, order, &now, &now],
    )?;

    let mut stmt = conn.prepare("SELECT * FROM paper_boards WHERE id = ?1")?;
    stmt.query_row([&id], |row| row_to_board(row))
}

pub fn update_board(
    conn: &Connection,
    id: &str,
    updates: &Value,
) -> rusqlite::Result<PaperBoard> {
    let mut sets = Vec::new();
    let mut values: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();

    if let Some(v) = updates.get("name") {
        sets.push("name = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("linkedNoteId") {
        sets.push("linked_note_id = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("viewportX") {
        sets.push("viewport_x = ?");
        values.push(Box::new(v.as_f64()));
    }
    if let Some(v) = updates.get("viewportY") {
        sets.push("viewport_y = ?");
        values.push(Box::new(v.as_f64()));
    }
    if let Some(v) = updates.get("viewportZoom") {
        sets.push("viewport_zoom = ?");
        values.push(Box::new(v.as_f64()));
    }
    if let Some(v) = updates.get("order") {
        sets.push("\"order\" = ?");
        values.push(Box::new(v.as_i64()));
    }

    if sets.is_empty() {
        let mut stmt = conn.prepare("SELECT * FROM paper_boards WHERE id = ?1")?;
        return stmt.query_row([id], |row| row_to_board(row));
    }

    sets.push("updated_at = datetime('now')");
    values.push(Box::new(id.to_string()));

    let sql = format!(
        "UPDATE paper_boards SET {} WHERE id = ?",
        sets.join(", ")
    );
    let params: Vec<&dyn rusqlite::types::ToSql> = values.iter().map(|v| v.as_ref()).collect();
    conn.execute(&sql, params.as_slice())?;

    let mut stmt = conn.prepare("SELECT * FROM paper_boards WHERE id = ?1")?;
    stmt.query_row([id], |row| row_to_board(row))
}

pub fn delete_board(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM paper_edges WHERE board_id = ?1", [id])?;
    conn.execute("DELETE FROM paper_nodes WHERE board_id = ?1", [id])?;
    conn.execute("DELETE FROM paper_boards WHERE id = ?1", [id])?;
    Ok(())
}

// --- Node functions ---

pub fn fetch_node_counts_by_board(conn: &Connection) -> rusqlite::Result<Value> {
    let mut stmt = conn.prepare(
        "SELECT board_id, COUNT(*) as count FROM paper_nodes GROUP BY board_id",
    )?;
    let mut map = serde_json::Map::new();
    stmt.query_map([], |row| {
        let board_id: String = row.get("board_id")?;
        let count: i64 = row.get("count")?;
        Ok((board_id, count))
    })?
    .for_each(|r| {
        if let Ok((board_id, count)) = r {
            map.insert(board_id, Value::from(count));
        }
    });
    Ok(Value::Object(map))
}

pub fn fetch_nodes_by_board(
    conn: &Connection,
    board_id: &str,
) -> rusqlite::Result<Vec<PaperNode>> {
    let mut stmt = conn.prepare(
        "SELECT * FROM paper_nodes WHERE board_id = ?1 \
         ORDER BY CASE WHEN parent_node_id IS NULL THEN 0 ELSE 1 END, z_index ASC",
    )?;
    let rows = stmt.query_map([board_id], |row| row_to_node(row))?;
    rows.collect()
}

pub fn create_node(
    conn: &Connection,
    params_val: &Value,
) -> rusqlite::Result<PaperNode> {
    let id = params_val
        .get("id")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| format!("node-{}", helpers::new_uuid()));
    let board_id = params_val.get("boardId").and_then(|v| v.as_str()).unwrap_or("");
    let node_type = params_val.get("nodeType").and_then(|v| v.as_str()).unwrap_or("default");
    let position_x = params_val.get("positionX").and_then(|v| v.as_f64()).unwrap_or(0.0);
    let position_y = params_val.get("positionY").and_then(|v| v.as_f64()).unwrap_or(0.0);
    let width = params_val.get("width").and_then(|v| v.as_f64()).unwrap_or(200.0);
    let height = params_val.get("height").and_then(|v| v.as_f64()).unwrap_or(100.0);
    let z_index = params_val.get("zIndex").and_then(|v| v.as_i64()).unwrap_or(0);
    let parent_node_id = params_val.get("parentNodeId").and_then(|v| v.as_str());
    let ref_entity_id = params_val.get("refEntityId").and_then(|v| v.as_str());
    let ref_entity_type = params_val.get("refEntityType").and_then(|v| v.as_str());
    let text_content = params_val.get("textContent").and_then(|v| v.as_str());
    let frame_color = params_val.get("frameColor").and_then(|v| v.as_str());
    let frame_label = params_val.get("frameLabel").and_then(|v| v.as_str());
    let label = params_val.get("label").and_then(|v| v.as_str());
    let hidden = params_val.get("hidden").and_then(|v| v.as_bool()).unwrap_or(false);
    let now = helpers::now();

    conn.execute(
        "INSERT INTO paper_nodes \
         (id, board_id, node_type, position_x, position_y, width, height, z_index, \
          parent_node_id, ref_entity_id, ref_entity_type, text_content, \
          frame_color, frame_label, label, hidden, created_at, updated_at) \
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16, ?17, ?18)",
        params![
            &id,
            board_id,
            node_type,
            position_x,
            position_y,
            width,
            height,
            z_index,
            parent_node_id,
            ref_entity_id,
            ref_entity_type,
            text_content,
            frame_color,
            frame_label,
            label,
            hidden as i64,
            &now,
            &now,
        ],
    )?;

    let mut stmt = conn.prepare("SELECT * FROM paper_nodes WHERE id = ?1")?;
    stmt.query_row([&id], |row| row_to_node(row))
}

pub fn update_node(
    conn: &Connection,
    id: &str,
    updates: &Value,
) -> rusqlite::Result<PaperNode> {
    let mut sets = Vec::new();
    let mut values: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();

    if let Some(v) = updates.get("nodeType") {
        sets.push("node_type = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("positionX") {
        sets.push("position_x = ?");
        values.push(Box::new(v.as_f64()));
    }
    if let Some(v) = updates.get("positionY") {
        sets.push("position_y = ?");
        values.push(Box::new(v.as_f64()));
    }
    if let Some(v) = updates.get("width") {
        sets.push("width = ?");
        values.push(Box::new(v.as_f64()));
    }
    if let Some(v) = updates.get("height") {
        sets.push("height = ?");
        values.push(Box::new(v.as_f64()));
    }
    if let Some(v) = updates.get("zIndex") {
        sets.push("z_index = ?");
        values.push(Box::new(v.as_i64()));
    }
    if let Some(v) = updates.get("parentNodeId") {
        sets.push("parent_node_id = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("refEntityId") {
        sets.push("ref_entity_id = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("refEntityType") {
        sets.push("ref_entity_type = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("textContent") {
        sets.push("text_content = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("frameColor") {
        sets.push("frame_color = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("frameLabel") {
        sets.push("frame_label = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("label") {
        sets.push("label = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("hidden") {
        sets.push("hidden = ?");
        values.push(Box::new(v.as_bool().map(|b| b as i64)));
    }

    if sets.is_empty() {
        let mut stmt = conn.prepare("SELECT * FROM paper_nodes WHERE id = ?1")?;
        return stmt.query_row([id], |row| row_to_node(row));
    }

    sets.push("updated_at = datetime('now')");
    values.push(Box::new(id.to_string()));

    let sql = format!(
        "UPDATE paper_nodes SET {} WHERE id = ?",
        sets.join(", ")
    );
    let params: Vec<&dyn rusqlite::types::ToSql> = values.iter().map(|v| v.as_ref()).collect();
    conn.execute(&sql, params.as_slice())?;

    let mut stmt = conn.prepare("SELECT * FROM paper_nodes WHERE id = ?1")?;
    stmt.query_row([id], |row| row_to_node(row))
}

pub fn bulk_update_positions(
    conn: &Connection,
    updates: &[Value],
) -> rusqlite::Result<()> {
    let tx = conn.unchecked_transaction()?;

    for update in updates {
        let id = update.get("id").and_then(|v| v.as_str()).unwrap_or("");
        let position_x = update.get("positionX").and_then(|v| v.as_f64()).unwrap_or(0.0);
        let position_y = update.get("positionY").and_then(|v| v.as_f64()).unwrap_or(0.0);

        tx.execute(
            "UPDATE paper_nodes SET position_x = ?1, position_y = ?2, \
             updated_at = datetime('now') WHERE id = ?3",
            params![position_x, position_y, id],
        )?;
    }

    tx.commit()
}

pub fn bulk_update_z_indices(
    conn: &Connection,
    updates: &[Value],
) -> rusqlite::Result<()> {
    let tx = conn.unchecked_transaction()?;

    for update in updates {
        let id = update.get("id").and_then(|v| v.as_str()).unwrap_or("");
        let z_index = update.get("zIndex").and_then(|v| v.as_i64()).unwrap_or(0);

        tx.execute(
            "UPDATE paper_nodes SET z_index = ?1, updated_at = datetime('now') WHERE id = ?2",
            params![z_index, id],
        )?;
    }

    tx.commit()
}

pub fn delete_node(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    conn.execute(
        "DELETE FROM paper_edges WHERE source_node_id = ?1 OR target_node_id = ?1",
        [id],
    )?;
    // Unparent children
    conn.execute(
        "UPDATE paper_nodes SET parent_node_id = NULL WHERE parent_node_id = ?1",
        [id],
    )?;
    conn.execute("DELETE FROM paper_nodes WHERE id = ?1", [id])?;
    Ok(())
}

// --- Edge functions ---

pub fn fetch_edges_by_board(
    conn: &Connection,
    board_id: &str,
) -> rusqlite::Result<Vec<PaperEdge>> {
    let mut stmt = conn.prepare(
        "SELECT * FROM paper_edges WHERE board_id = ?1 ORDER BY created_at",
    )?;
    let rows = stmt.query_map([board_id], |row| row_to_edge(row))?;
    rows.collect()
}

pub fn create_edge(
    conn: &Connection,
    params_val: &Value,
) -> rusqlite::Result<PaperEdge> {
    let id = params_val
        .get("id")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| format!("edge-{}", helpers::new_uuid()));
    let board_id = params_val.get("boardId").and_then(|v| v.as_str()).unwrap_or("");
    let source_node_id = params_val.get("sourceNodeId").and_then(|v| v.as_str()).unwrap_or("");
    let target_node_id = params_val.get("targetNodeId").and_then(|v| v.as_str()).unwrap_or("");
    let source_handle = params_val.get("sourceHandle").and_then(|v| v.as_str());
    let target_handle = params_val.get("targetHandle").and_then(|v| v.as_str());
    let label = params_val.get("label").and_then(|v| v.as_str());
    let style_json = params_val.get("styleJson").and_then(|v| v.as_str());
    let now = helpers::now();

    conn.execute(
        "INSERT INTO paper_edges \
         (id, board_id, source_node_id, target_node_id, source_handle, \
          target_handle, label, style_json, created_at) \
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)",
        params![
            &id,
            board_id,
            source_node_id,
            target_node_id,
            source_handle,
            target_handle,
            label,
            style_json,
            &now,
        ],
    )?;

    let mut stmt = conn.prepare("SELECT * FROM paper_edges WHERE id = ?1")?;
    stmt.query_row([&id], |row| row_to_edge(row))
}

pub fn delete_edge(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM paper_edges WHERE id = ?1", [id])?;
    Ok(())
}
